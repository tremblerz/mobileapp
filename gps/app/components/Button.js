import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';
import * as React from 'react';
import { SvgXml } from 'react-native-svg';

import { Typography } from './Typography';

/**
 * @param {{
 *   label: string;
 *   secondary?: boolean;
 *   icon?: string;
 *   onPress: () => void;
 *   disabled?: boolean;
 * }} param0
 */
export const Button = ({
  label,
  secondary,
  icon,
  onPress,
  disabled,
  small,
  ...otherProps
}) => {
  return (
    <Container
      onPress={onPress}
      accessible
      accessibilityLabel={label}
      accessibilityRole='button'
      hasIcon={!!icon}
      secondary={secondary}
      disabled={disabled}
      small={small}
      {...otherProps}>
      <ThemeProvider theme={invertTextColors}>
        <Label small={small} secondary={secondary} disabled={disabled}>
          {label}
        </Label>
        {icon ? <Icon xml={icon} /> : null}
      </ThemeProvider>
    </Container>
  );
};

const invertTextColors = theme => {
  return {
    ...theme,
    textPrimaryOnBackground: theme.onPrimary,
    textSecondaryOnBackground: theme.primary,
  };
};

const themePrimary = ({ theme, secondary, disabled }) =>
  disabled ? theme.disabled : secondary ? 'transparent' : theme.primary;

const getBorderColor = ({ theme, secondary }) =>
  secondary ? theme.primary : 'transparent';

const getJustifyContent = ({ hasIcon }) =>
  hasIcon ? 'space-between' : 'center';

const Container = styled.TouchableOpacity`
  background-color: ${themePrimary};
  height: ${({ small }) => (small ? '48px' : '72px')};
  border: 2px solid ${getBorderColor};
  padding-horizontal: ${({ small }) => (small ? '14px' : '16px')};
  border-radius: 8px;
  flex-direction: row;
  justify-content: ${getJustifyContent};
  align-content: center;
  align-items: center;
  align-self: stretch;
`;

const getFontSize = ({ small }) => (small ? '14px' : '20px');

const getLineHeight = ({ small }) => (small ? '21px' : '40px');

const Label = styled(Typography)`
  font-family: IBMPlexSans-Medium;
  font-size: ${getFontSize};
  line-height: ${getLineHeight};
  font-weight: normal;
`;

const Icon = styled(SvgXml)`
  margin-right: 8px; // for visual alignment
`;
