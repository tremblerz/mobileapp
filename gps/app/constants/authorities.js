export const PUBLIC_DATA_URL =
  'https://raw.githubusercontent.com/beoutbreakprepared/nCoV2019/master/latest_data/latestdata.csv';

export const AUTHORITIES_LIST_URL =
  'https://raw.githubusercontent.com/Path-Check/safeplaces-frontend/develop/healthcare-authorities.yaml';
