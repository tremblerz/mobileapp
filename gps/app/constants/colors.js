const colors = {
  BLACK: '#000',
  WHITE: '#FFF',
  DODGER_BLUE: '#428AF8',
  SILVER: '#BEBEBE',
  TORCH_RED: '#F8262F',
  MISCHKA: '#E5E4E6',
  PRIMARY_TEXT: '#000',
  SUCCESS: '#41dca4', // Green
  WARNING: '#ffc000', // Orange
  VIOLET_ALPHA_06: '#f7f8ff', // cannot use transparency
  VIOLET: '#4051DB',
  TRANSPARENT: 'transparent',

  // For notifications
  HIGHEST_RISK: '#e74c3c',
  MIDDLE_RISK: '#e74c3c',
  LOWER_RISK: '#e74c3c',
  LOWEST_RISK: '#e74c3c',

  REG_BUTTON: '#428AF8',
  POS_BUTTON: '#32A852',
  NEG_BUTTON: '#F8262F',
  SENS_BUTTON: '#6C3794',
  VIOLET_BUTTON: '#6979F8',
  VIOLET_BUTTON_DARK: '#3A4CD7',

  VIOLET_TEXT: '#3C4ED8',

  GRAY_BACKGROUND: '#DADADA',

  INTRO_WHITE_BG: '#F7F8FF',
  PULSE_WHITE: '#FFFFFF40',

  DIVIDER: '#A5AFFB',

  BLUE_BUTTON: '#05369B',
  BLUE_TO_BUTTON: '#185BD3',
  RED_BUTTON: '#E3554E',
  RED_TO_BUTTON: '#E97C79',
  GRAY_BUTTON: '#868686',
  GRAY_TO_BUTTON: '#868686',

  MAP_LINE_STROKE: '#4052DB',

  BOTTOM_SHEET_HEADER_INDICATOR: '#E1E4FF',

  BLUE_LINK: '#007AFF',

  RED_TEXT: '#FF5656',

  NAV_BAR_VIOLET: '#5061E6',
  NAV_BAR_BORDER: 'rgba(189, 195, 199,0.6)',

  DARK_GRAY: '#3C475B',
  LIGHT_GRAY: '#EEEEEE',

  MONO_DARK: '#202020',
  MONO_SECONDARY: '#757677',
};

export default colors;
