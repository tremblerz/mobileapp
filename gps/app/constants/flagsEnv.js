import env from 'react-native-config';

/**
 * Normalizes flags:
 *
 * Example:
 *
 * `{ flag_a: 'true' }` becomes `{ a: true }`
 *
 * @param {{[key: string]: string}} envConfig
 */
export function parseFlags(envConfig) {
  return Object.entries(envConfig)
    .filter(([key]) => key.toLowerCase().startsWith('flag'))
    .reduce((flags, [key, value]) => {
      const flag = key.replace(/^flag_/i, '');

      flags[flag] = value === 'true' || value === '1';
      return flags;
    }, {});
}

export const getBuildtimeFlags = () => parseFlags(env);
