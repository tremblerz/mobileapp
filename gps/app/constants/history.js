/**
 * Max esposure reporting window in days
 */
export const MAX_EXPOSURE_WINDOW_DAYS = 14;

/**
 * The value in minutes of each "bin" in the crossed path data.
 */
export const DEFAULT_EXPOSURE_PERIOD_MINUTES = 5;

/**
 * The value in minutes of how long an exposure at a location is
 *    considered concerning.
 */
export const CONCERN_TIME_WINDOW_MINUTES = 4 * 60; // 4 hours, in minutes

/**
 * The value in minutes of how frequently we should check intersection data if
 *    there has been no change to the authorities
 */
export const MIN_CHECK_INTERSECT_INTERVAL = 6 * 60;

/**
 * The value in minutes for the background task service to register for firing
 */
export const INTERSECT_INTERVAL = 60 * 12; // 12 Hours, the value is received in minutes

/**
 * Format of a single history item
 *
 * @typedef {{
 *   date: import("dayjs").Dayjs,
 *   exposureMinutes: number,
 * }} HistoryDay
 */

/**
 * Exposure history
 *
 * @typedef {!HistoryDay[]} History
 */
