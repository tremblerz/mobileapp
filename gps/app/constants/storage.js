export const LOCATION_DATA = 'LOCATION_DATA';
export const CONTACT_DATA = 'CONTACT_DATA';
export const PARTICIPATE = 'PARTICIPATE';
export const MY_UUIDs = 'MY_UUIDs';
export const CROSSED_PATHS = 'CROSSED_PATHS';
export const LANG_OVERRIDE = 'LANG_OVERRIDE';
export const DEBUG_MODE = 'DEBUG_MODE';
export const AUTHORITY_NEWS = 'AUTHORITY_NEWS';
export const LAST_CHECKED = 'LAST_CHECKED';
export const AUTHORITY_SOURCE_SETTINGS = 'AUTHORITY_SOURCE_SETTINGS';
export const ENABLE_HCA_AUTO_SUBSCRIPTION = 'ENABLE_HCA_AUTO_SUBSCRIPTION';
export const FEATURE_FLAG_VALS = 'FEATURE_FLAG_VALS';
