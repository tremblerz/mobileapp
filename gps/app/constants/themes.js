import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';

import Color from './colors';

/** Violet on pale violet bg. e.g. Settings */
export const defaultTheme = {
  background: Color.VIOLET_ALPHA_06,
  textPrimaryOnBackground: Color.VIOLET,
  textSecondaryOnBackground: 'rgba(64, 81, 219, 0.6)',

  navBar: Color.VIOLET,
  onNavBar: Color.WHITE,

  /** E.g. button bg */
  primary: Color.VIOLET_BUTTON,
  /** E.g. button text color */
  onPrimary: Color.WHITE,

  success: Color.SUCCESS,
  warning: Color.WARNING,
  disabled: 'rgba(64, 81, 219, 0.6)',
  border: Color.DIVIDER,
};

/** White on violet bg. E.g. Main screen */
export const violet = {
  ...defaultTheme,
  background: `linear-gradient(245.21deg, ${Color.VIOLET_BUTTON} 0%, ${Color.VIOLET_BUTTON_DARK} 100%);`,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: 'rgba(255, 255, 255, 0.6)',

  navBarBorder: Color.VIOLET, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.VIOLET,
  disabled: '#ececec',
};

/** White on gray bg. E.g. Possible exposure mode on default screen */
export const charcoal = {
  ...defaultTheme,
  background: Color.DARK_GRAY,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: Color.LIGHT_GRAY,

  navBar: Color.DARK_GRAY,
  navBarBorder: Color.DARK_GRAY, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.DARK_GRAY,

  disabled: '#A9AFBA',
};

const THEME_MAP = {
  default: defaultTheme,
  violet,
  charcoal,
};

/** A view which uses the theme's background color */
export const ThemedBackground = styled.View`
  background-color: ${({ theme }) => theme.background};
`;

/**
 * Return theme config for a known theme name
 *
 * @param {'default' | 'violet' | 'charcoal'} themeName
 */
export const getTheme = themeName => {
  const themeColors = THEME_MAP[themeName];
  if (!themeColors) {
    throw new Error(`Unknown theme ${themeName}`);
  }
  return themeColors;
};

/**
 * Configures themeable sub-components with background, text (primary,
 * secondary), button and other various theme colors.
 *
 * Usage:
 *
 * This will render a charcoal view, with a white heading, and a grey paragraph.
 *
 * ```jsx
 * <Theme use="charcoal" setBackground>
 *   <Typography use="headline2">Test</Typogrphy>
 *   <Typography secondary>Paragraph here</Typogrphy>
 * </Theme>
 * ```
 *
 * It makes available the following `props.theme.x` properties in styled components:
 *
 * - background - the standard bg color
 * - primary - the primary theme/button color on top of bakground (e.g. violet)
 * - onPrimary - color to show on top of primary (usually white)
 * - onPrimary - color to show on top of primary (usually white)
 * - textPrimaryOnBackground - text color to show on background
 * - textSecondaryOnBackground - secondary text color to show on background
 * - warning - warning color (always orange)
 * - success - success color (always green)
 * - border - border/divider color (usually gray)
 *
 * @param {{
 *   use?: string,
 *   setBackground?: boolean,
 * }} param0
 */
export const Theme = ({ use = 'default', setBackground = false, children }) => {
  return (
    <ThemeProvider theme={getTheme(use)}>
      {setBackground ? (
        <ThemedBackground>{children}</ThemedBackground>
      ) : (
        children
      )}
    </ThemeProvider>
  );
};

/**
 * @param {{theme: defaultTheme, secondary?: boolean}} param0
 */
export const themeTextColor = ({ theme, secondary }) =>
  secondary ? theme.textSecondaryOnBackground : theme.textPrimaryOnBackground;
