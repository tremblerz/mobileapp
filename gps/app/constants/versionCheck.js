export const APPSTORE_URL = 'itms-apps://itunes.apple.com/us/app/id1508266966';
export const PLAYSTORE_URL = 'market://details?id=org.pathcheck.covidsafepaths';
export const VERSION_URL =
  'https://rawcdn.githack.com/Path-Check/covid-safe-paths/master/versions.yaml';
export const YAML_MANDATORY_VERSION_KEY = 'mandatory upgrade version';
export const YAML_LATEST_VERSION_KEY = 'upgrade version';
