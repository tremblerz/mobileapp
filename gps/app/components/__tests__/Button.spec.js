import { render } from '@testing-library/react-native';
import React from 'react';

import { Theme } from '../../constants/themes';
import { Button } from '../Button';

const svg = `<g />`;

it('renders a violet 72px button on default theme', () => {
  const { asJSON } = render(
    <Theme>
      <Button label='Default' />
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('allows small override', () => {
  const { asJSON } = render(
    <Theme>
      <Button small label='Small' />
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('renders different color for secondary', () => {
  const { asJSON } = render(
    <Theme>
      <Button label='Button' secondary />
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('renders an icon and space-between if supplied', () => {
  const { asJSON } = render(
    <Theme>
      <Button label='Button' icon={svg} />
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('changes color based on theme', () => {
  const { asJSON } = render(
    <>
      <Theme use='default'>
        <Button label='onDefault' />
      </Theme>
      <Theme use='charcoal'>
        <Button label='onCharcoal' />
      </Theme>
      <Theme use='violet'>
        <Button label='onViolet' />
      </Theme>
    </>,
  );

  expect(asJSON()).toMatchSnapshot();
});
