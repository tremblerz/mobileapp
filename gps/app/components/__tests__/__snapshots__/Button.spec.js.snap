// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows small override 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityLabel="Small"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#6979F8",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 14,
      }
    }
  >
    <Text
      small={true}
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 14,
          "fontWeight": "normal",
          "lineHeight": 21,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      Small
    </Text>
  </View>
</View>
`;

exports[`changes color based on theme 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityLabel="onDefault"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#6979F8",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      onDefault
    </Text>
  </View>
  <View
    accessibilityLabel="onCharcoal"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#FFF",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3C475B",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      onCharcoal
    </Text>
  </View>
  <View
    accessibilityLabel="onViolet"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#FFF",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#4051DB",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      onViolet
    </Text>
  </View>
</View>
`;

exports[`renders a violet 72px button on default theme 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityLabel="Default"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#6979F8",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      Default
    </Text>
  </View>
</View>
`;

exports[`renders an icon and space-between if supplied 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityLabel="Button"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#6979F8",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "transparent",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "space-between",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      Button
    </Text>
    <SvgXml
      style={
        Object {
          "marginRight": 8,
        }
      }
      xml="<g />"
    />
  </View>
</View>
`;

exports[`renders different color for secondary 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityLabel="Button"
    accessibilityRole="button"
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    style={
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "transparent",
        "borderBottomLeftRadius": 8,
        "borderBottomRightRadius": 8,
        "borderColor": "#6979F8",
        "borderStyle": "solid",
        "borderTopLeftRadius": 8,
        "borderTopRightRadius": 8,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "center",
        "opacity": 1,
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      secondary={true}
      style={
        Object {
          "color": "#6979F8",
          "fontFamily": "IBMPlexSans-Medium",
          "fontSize": 20,
          "fontWeight": "normal",
          "lineHeight": 40,
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      Button
    </Text>
  </View>
</View>
`;
